zerver/lib/upload.py

```diff
@@ -573,6 +582,29 @@ def get_emoji_url(self, emoji_file_name: str, realm_id: int) -> str:
                                                         emoji_file_name=emoji_file_name)
         return "https://%s.s3.amazonaws.com/%s" % (bucket, emoji_path)
 
+    def upload_export_tarball(self, realm: Optional[Realm], tarball_path: str) -> str:
+        def percent_callback(complete: Any, total: Any) -> None:
+            sys.stdout.write('.')
+            sys.stdout.flush()
+
+        conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
+        # We use the avatar bucket, because it's world-readable.
+        bucket = get_bucket(conn, settings.S3_AVATAR_BUCKET)
+        key = Key(bucket)
+        key.key = os.path.join("exports", generate_random_token(32), os.path.basename(tarball_path))
+        key.set_contents_from_filename(tarball_path, cb=percent_callback, num_cb=40)
+
+        public_url = 'https://{bucket}.{host}/{key}'.format(
+            host=conn.server_name(),
+            bucket=bucket.name,
+            key=key.key)
+        return public_url
+
+    def delete_export_tarball(self, uri: str) -> bool:
+        path_id = re.sub('https://test-avatar-bucket.s3.amazonaws.com:443/', '', uri)
```

Comment:
I don't understand this `re.sub` logic; how is it supposed to work?  

I would have expected code involving `urllib.parse` if the plan is to parse a `uri`...

Reply:
(It's also important here that the unit tests can verify the logic for what file we're deleting is correct; the way you have things setup it's a bit unclear to me how this is supposed to work)

Reply:
Basically I was thinking that this just reverse engineers the key from
```
key.key = os.path.join("exports", generate_random_token(32), os.path.basename(tarball_path))
```
When the upload is created in `upload_export_tarball` as this returns a string in the form:
```
exports/4f256889b77c912806eadc1fc28aa813/zulip-export-8vwxgm9r.tar.gz
``` 
 Although, I see how this is a weird and how using `urllib.parse` would be much more appropriate.  But is this sort of approach of grabbing the key from the uri acceptable as a means to delete the file?

I also see how one couldn't verify the correct file is being deleted, I'll rethink this with the tests validating the file.

Reply:
I think the updated approach now addresses this issue:
> (It's also important here that the unit tests can verify the logic for what file we're deleting is correct; the way you have things setup it's a bit unclear to me how this is supposed to work)

(I've also been going off of this comment for reference: https://github.com/zulip/zulip/pull/11996#discussion_r287917477)